src/types/chat.rs

```diff
@@ -61,108 +66,124 @@ impl ChatJson {
-                            if let Some(&Json::String(ref string)) = event.get("action") {
-                                result.hover_event = match string.as_slice() {
-                                    "show_text" => Some(HoverEvent::Text(val)),
-                                    "show_achievement" => Some(HoverEvent::Achievement(val)),
-                                    "show_item" => Some(HoverEvent::Item(val)),
+                        // Handle the JSON format of hover events.
+                        "hoverEvent" => {
+                            if let Json::Object(event) = value {
+                                // Get the `value` first.
+                                let val: String = match event.get("value") {
+                                    Some(&Json::String(ref string)) => string.clone(),
                                     _ => return Err(ChatJsonError::InvalidHoverEvent)
                                 };
+                                // Handle the different click events.
+                                if let Some(&Json::String(ref string)) = event.get("action") {
+                                    result.hover_event = match string.as_slice() {
+                                        "show_text" => Some(HoverEvent::Text(val)),
+                                        "show_achievement" => Some(HoverEvent::Achievement(val)),
+                                        "show_item" => Some(HoverEvent::Item(val)),
+                                        _ => return Err(ChatJsonError::InvalidHoverEvent)
+                                    };
+                                } else {
+                                    return Err(ChatJsonError::InvalidHoverEvent);
+                                }
                             } else {
-                                return Err(ChatJsonError::InvalidHoverEvent);
+                                return Err(ChatJsonError::InvalidFieldType);
                             }
                         }
-                    },
-                    "extra" => {
-                        if let Json::Array(extra) = value {
-                            result.extra = Some(extra);
-                        } else {
-                            return Err(ChatJsonError::InvalidFieldType);
+                        "extra" => {
+                            if let Json::Array(extra) = value {
+                                result.extra = try!(extra.into_iter().map(|elt| ChatJson::from_json(elt)).collect());
+                            } else {
+                                return Err(ChatJsonError::InvalidFieldType);
+                            }
                         }
-                    },
-                    // TODO: Error on unknown key when the implementation is complete.
-                    _ => (),
-                };
+                        // TODO: Error on unknown key when the implementation is complete.
+                        _ => ()
```

Comment:
Oops. I missed this. I think that changing this to `v => Err(ChatJsonError::UnknownField(v.to_string()),` should be enough to account for unknown fields.